homeassistant/components/emulated_hue/hue_api.py

```diff
@@ -393,141 +459,66 @@ def __init__(self, config):
                 hass.services.async_call(domain, service, data, blocking=True)
             )
 
+        # Create success responses for all received keys
         json_response = [
             create_hue_success_response(entity_id, HUE_API_STATE_ON, parsed[STATE_ON])
         ]
 
-        if parsed[STATE_BRIGHTNESS] is not None:
-            json_response.append(
-                create_hue_success_response(
-                    entity_id, HUE_API_STATE_BRI, parsed[STATE_BRIGHTNESS]
-                )
-            )
-        if parsed[STATE_HUE] is not None:
-            json_response.append(
-                create_hue_success_response(
-                    entity_id, HUE_API_STATE_HUE, parsed[STATE_HUE]
+        for (key, val) in (
+            # (STATE_ON, HUE_API_STATE_ON),
```

Comment:
(nitpick) Commented out code, I think it can be removed (returning the state happens in line 464)

Reply:
Thanks for the review. Doing the best I can, it's kind of a messy implementation. Hopefully one day we'll get it all working perfectly but yeah that's my standpoint too - for now, get it *mostly* working (and fix the big issues lots of people are seeing) and go back and add the finishing touches as we can.

> Thanks @NobleKangaroo for putting together this PR!
> 
> I've been looking at the problems in this component over the last few days too. All of the changes here make sense, and I'm pleased to report that this fixes the majority of the issues I've been seeing.
> 
> I am still able to trigger occasional glitches such as 'Device malfunction' in an MPD media player, but at this stage I'd rather that this PR gets merged as it is, and work on any remaining issues once all the other fixes are in place. Judging by the forum comments and open issues, I think it would fix the majority of the problems people are having.

Personally I've yet to see a single "Device malfunction" or "Value out of range" error since implementing this using two LIFX bulbs and both an MQTT fan and wax warmer (Tasmota dual-outlet plugs), but it's definitely possible there's still edge cases we've not yet uncovered. Harkens back to the "go back and tough things up" sentiment. ;)

> The only thing that I'm not 100% convinced on is the change of entity id from string to name. I will miss the ability to read the entity name straight from the API, but I guess its not a big issue because the Name attribute is still there so the device name can be read. I understand that other people have been having problems with the current scheme.

Yeah it's not the most user-friendly but to be honest, it's Hue-compliant so in the end I would think it's probably better for the system to adhere to the API as closely as possible. Can always rename the entity yourself; in the end, the id just gets hashed and colon-delimited before being passed to Alexa and Google so outside of that (and in emulated_hue_ids.json of course), there shouldn't be much of a change. I adapted the code to this simple script, for generating the IDs faster (for tests, etc):
```
# Usage: python id.py deviceid
import hashlib
import sys

unique_id = hashlib.md5(sys.argv[1].encode()).hexdigest()
unique_id = "00:%s:%s:%s:%s:%s:%s:%s-%s" % (
    unique_id[0:2],
    unique_id[2:4],
    unique_id[4:6],
    unique_id[6:8],
    unique_id[8:10],
    unique_id[10:12],
    unique_id[12:14],
    unique_id[14:16],
)
print(unique_id)
```

> (nitpick) Commented out code, I think it can be removed (returning the state happens in line 464)

Absolutely. I had actually tried to create an empty json response and populate it with state_on ONLY if the device was being turned on but that caused issues. IIRC, in short: for devices that are off, when you set them to a color using the Alexa app, they ALSO need to reply with a "state: on" message or else they don't do anything and Alexa reports that they're malfunctioning. I left the code in there JUST IN CASE someone wanted to take another stab at this. I can remove it if that's preferred - no big deal!

Hoping to work this all out. I feel we're getting closer, that's for sure!

Reply:
> it's kind of a messy implementation

Yeah, well l think we're moving in the right direction as long as each update makes it slightly less messy. And there are several pieces here that are a big improvement already, thank you!

> Personally I've yet to see a single "Device malfunction" or "Value out of range" error since implementing this

I was seeing errors in quite a few places beforehand, and they are almost completely gone now. I've got a development setup where I can see if I can retrigger it and capture the message sequence. I'll report back if I can work out what's going on.

> Harkens back to the "go back and tough things up" sentiment. ;)

Yep :muscle: 

> Yeah it's not the most user-friendly but to be honest, it's Hue-compliant so in the end I would think it's probably better for the system to adhere to the API as closely as possible. 

Yes, agreed.

> Absolutely. I had actually tried to create an empty json response and populate it with state_on ONLY if the device was being turned on but that caused issues. IIRC, in short: for devices that are off, when you set them to a color using the Alexa app, they ALSO need to reply with a "state: on" message or else they don't do anything and Alexa reports that they're malfunctioning.

I think I found that not sending the state at all is okay in that case - did you try that?

> I left the code in there JUST IN CASE someone wanted to take another stab at this. I can remove it if that's preferred - no big deal!

Might be better to open an issue that describes the problem and how far you got with it?

> Hoping to work this all out. I feel we're getting closer, that's for sure!

:+1: